File: code/sorting/merge_sort/merge_sort.scala
object Main
{

  def merge(left: List[Int], right: List[Int]): List[Int] =
  {
    if (left.isEmpty)
    {
      return right
    }
    else if (right.isEmpty)
    {
      return left
    }

    if (left.head < right.head)
    {
      return left.head :: merge(left.tail, right)
    }
    else
    {
      return right.head :: merge(left, right.tail)
    }
  }

  def merge_sort(xs: List[Int]): List[Int] = {
    if (xs.size == 1) {
      return xs
    }

    merge(merge_sort(xs.slice(0, xs.size/2)), merge_sort(xs.slice(xs.size/2, xs.size)))
  }

  def main(args: Array[String]) =
  {
    println(merge(List(1,3,5), List(2,4,6)))
    println(merge_sort(List(4,2,3,1,6,2,7,1,116,3,11,135)))
  }
}

File: code/sorting/src/merge_sort/merge_sort.js
var a = [34, 203, 3, 746, 200, 984, 198, 764, 9];
// Part of Cosmos by OpenGenus Foundation
function mergeSort(arr) {
  if (arr.length < 2) return arr;

  var middle = parseInt(arr.length / 2);
  var left = arr.slice(0, middle);
  var right = arr.slice(middle, arr.length);

  return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
  var result = [];

  while (left.length && right.length) {
    if (left[0] <= right[0]) {
      result.push(left.shift());
    } else {
      result.push(right.shift());
    }
  }

  while (left.length) result.push(left.shift());

  while (right.length) result.push(right.shift());

  return result;
}

console.log(mergeSort(a));
